### PR TITLE
Homework 1 completed.

### PR DESCRIPTION
**Pull request by kdr1mkf**
Homework 1 seems to be completed.

### CODE REVIEW

File: src/test/java/com/epam/tc/hw1/function/multiply/MultiplyFunctionTest.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package com.epam.tc.hw1.function.multiply;
+
+import com.epam.tc.hw1.CommonTest;
+import org.testng.annotations.Test;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+public class MultiplyFunctionTest extends CommonTest {
+
+    final String failedTestMessage = "Multiply function test failed. A %s input has been provided.";
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/function/multiply/MultiplyFunctionTest.java:9]: lets make this variable private final static

File: src/test/java/com/epam/tc/hw1/function/multiply/MultiplyFunctionTest.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Test(groups = "positive integer input",
+          dataProviderClass = MultiplyFunctionDataProvider.class,
+          dataProvider = "positive integers")
+    public void testPositiveIntInput(double multiplicand, double multiplier, double expectedResult) {
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/function/multiply/MultiplyFunctionTest.java:14]: test should be the last word in method name.
also, arguments are doubles, but the method name has 'int' in name.
lets rename this method to multiplyPositiveNumbersTest

File: src/test/java/com/epam/tc/hw1/function/subtract/SubtractFunctionTest.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+public class SubtractFunctionTest extends CommonTest {
+
+    final String failedTestMessage = "Subtract function test failed. A %s input has been provided.";
+
+    @Test(groups = "positive integer input",
+          dataProviderClass = SubtractFunctionDataProvider.class,
+          dataProvider = "positive integers")
+    public void testPositiveIntInput(double minuend, double subtrahend, double expectedResult) {
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/function/subtract/SubtractFunctionTest.java:14]: test should be the last word in method name.
also, arguments are doubles, but the method name has 'int' in name.
lets rename this method to substractPositiveNumbersTest

**Review by ZeroShadows — COMMENTED** [review]
general comments - lets remove .gitkeep files from not empthy folders

File: src/test/java/com/epam/tc/hw1/function/add/AddFunctionTest.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
+package com.epam.tc.hw1.function.add;
+
+import com.epam.tc.hw1.CommonTest;
+import org.testng.annotations.Test;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+public class AddFunctionTest extends CommonTest {
+
+    final String failedTestMessage = "Add function test failed. A %s input has been provided.";
```
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/add/AddFunctionTest.java]: String var changed into private static final.

File: src/test/java/com/epam/tc/hw1/function/divide/DivideFunctionTest.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
+package com.epam.tc.hw1.function.divide;
+
+import com.epam.tc.hw1.CommonTest;
+import org.testng.annotations.Test;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+import static org.assertj.core.api.Assertions.offset;
+
+public class DivideFunctionTest extends CommonTest {
+
+    final String failedTestMessage = "Divide function test failed. A %s input has been provided.";
```
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/divide/DivideFunctionTest.java]: String var changed into private static final.

File: src/test/java/com/epam/tc/hw1/function/multiply/MultiplyFunctionTest.java (hunk 1 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/multiply/MultiplyFunctionTest.java]: String var changed into private static final.

File: src/test/java/com/epam/tc/hw1/function/subtract/SubtractFunctionTest.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
+package com.epam.tc.hw1.function.subtract;
+
+import com.epam.tc.hw1.CommonTest;
+import org.testng.annotations.Test;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+public class SubtractFunctionTest extends CommonTest {
+
+    final String failedTestMessage = "Subtract function test failed. A %s input has been provided.";
```
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/subtract/SubtractFunctionTest.java]: String var changed into private static final.

File: src/test/java/com/epam/tc/hw1/function/add/AddFunctionTest.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Test(groups = "positive integer input",
+          dataProviderClass = AddFunctionDataProvider.class,
+          dataProvider = "positive integers")
+    public void testPositiveIntInput(double augend, double addend, double expectedResult) {
```
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/add/AddFunctionTest.java]: Input data types are changed (double -> long). See commit below.

As for methods naming, I would rather leave names as is. The code convention says, that a mothod's name should be a verb (https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html). "Test positive int input" is an action (is a verb). "Positive int input test" or "sumPositiveNumbersTest" is not an action (is a noun).

BTW, I have checked tests on epam autocode service. Test methods have names like: testSingleRoot, test2Roots, etc.

Here is another reason why I wouldn't change the method's name into sumPositiveNumbersTest. The mothod is contained by the class AddFunctionTest. I think no need to repeat the function name in the method's name.

File: src/test/java/com/epam/tc/hw1/function/add/AddFunctionTest.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        double actualResult = calculator.sum(augend, addend);
+        assertThat(actualResult)
+            .as(String.format(failedTestMessage, "positive integer"))
+            .isEqualTo(expectedResult);
+    }
+
+    @Test(groups = "negative integer input",
+          dataProviderClass = AddFunctionDataProvider.class,
+          dataProvider = "negative integers")
+    public void testNegativeIntInput(double augend, double addend, double expectedResult) {
```
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/add/AddFunctionTest.java]: Same comment.

Input data types are changed (double -> long). See commit below.

As for methods naming, I would rather leave names as is. The code convention says, that a mothod's name should be a verb (https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html). "Test positive int input" is an action (is a verb). "Positive int input test" or "sumPositiveNumbersTest" is not an action (is a noun).

BTW, I have checked tests on epam autocode service. Test methods have names like: testSingleRoot, test2Roots, etc.

Here is another reason why I wouldn't change the method's name into sumPositiveNumbersTest. The mothod is contained by the class AddFunctionTest. I think no need to repeat the function name in the method's name

File: src/test/java/com/epam/tc/hw1/function/add/AddFunctionTest.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        double actualResult = calculator.sum(augend, addend);
+        assertThat(actualResult)
+            .as(String.format(failedTestMessage, "negative integer"))
+            .isEqualTo(expectedResult);
+    }
+
+    @Test(groups = "mixed integer input",
+          dataProviderClass = AddFunctionDataProvider.class,
+          dataProvider = "mixed integers")
+    public void testMixedIntInput(double augend, double addend, double expectedResult) {
```
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/add/AddFunctionTest.java]: Same comment.

Input data types are changed (double -> long). See commit below.

As for methods naming, I would rather leave names as is. The code convention says, that a mothod's name should be a verb (https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html). "Test positive int input" is an action (is a verb). "Positive int input test" or "sumPositiveNumbersTest" is not an action (is a noun).

BTW, I have checked tests on epam autocode service. Test methods have names like: testSingleRoot, test2Roots, etc.

Here is another reason why I wouldn't change the method's name into sumPositiveNumbersTest. The mothod is contained by the class AddFunctionTest. I think no need to repeat the function name in the method's name

File: src/test/java/com/epam/tc/hw1/function/divide/DivideFunctionTest.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Test(groups = "positive integer input",
+          dataProviderClass = DivideFunctionDataProvider.class,
+          dataProvider = "positive integers")
+    public void testPositiveIntInput(double dividend, double divisor, double expectedResult) {
```
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/divide/DivideFunctionTest.java]: Input data types are changed (double -> long). See commit below.

As for methods naming, I would rather leave names as is. The code convention says, that a mothod's name should be a verb (https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html). "Test positive int input" is an action (is a verb). "Positive int input test" or "sumPositiveNumbersTest" is not an action (is a noun).

BTW, I have checked tests on epam autocode service. Test methods have names like: testSingleRoot, test2Roots, etc.

Here is another reason why I wouldn't change the method's name into dividePositiveNumbersTest. The mothod is contained by the class DivideFunctionTest. I think no need to repeat the function name in the method's name

File: src/test/java/com/epam/tc/hw1/function/divide/DivideFunctionTest.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        double actualResult = calculator.div(dividend, divisor);
+        assertThat(actualResult)
+            .as(String.format(failedTestMessage, "positive integer"))
+            .isEqualTo(expectedResult, offset(delta));
+    }
+
+    @Test(groups = "negative integer input",
+          dataProviderClass = DivideFunctionDataProvider.class,
+          dataProvider = "negative integers")
+    public void testNegativeIntInput(double dividend, double divisor, double expectedResult) {
```
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/divide/DivideFunctionTest.java]: Same comment.

Input data types are changed (double -> long). See commit below.

As for methods naming, I would rather leave names as is. The code convention says, that a mothod's name should be a verb (https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html). "Test positive int input" is an action (is a verb). "Positive int input test" or "sumPositiveNumbersTest" is not an action (is a noun).

BTW, I have checked tests on epam autocode service. Test methods have names like: testSingleRoot, test2Roots, etc.

Here is another reason why I wouldn't change the method's name into dividePositiveNumbersTest. The mothod is contained by the class DivideFunctionTest. I think no need to repeat the function name in the method's name

File: src/test/java/com/epam/tc/hw1/function/divide/DivideFunctionTest.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+        double actualResult = calculator.div(dividend, divisor);
+        assertThat(actualResult)
+            .as(String.format(failedTestMessage, "negative integer"))
+            .isEqualTo(expectedResult, offset(delta));
+    }
+
+    @Test(groups = "mixed integer input",
+          dataProviderClass = DivideFunctionDataProvider.class,
+          dataProvider = "mixed integers")
+    public void testMixedIntInput(double dividend, double divisor, double expectedResult) {
```
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/divide/DivideFunctionTest.java]: Same comment.

Input data types are changed (double -> long). See commit below.

As for methods naming, I would rather leave names as is. The code convention says, that a mothod's name should be a verb (https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html). "Test positive int input" is an action (is a verb). "Positive int input test" or "sumPositiveNumbersTest" is not an action (is a noun).

BTW, I have checked tests on epam autocode service. Test methods have names like: testSingleRoot, test2Roots, etc.

Here is another reason why I wouldn't change the method's name into dividePositiveNumbersTest. The mothod is contained by the class DivideFunctionTest. I think no need to repeat the function name in the method's name

File: src/test/java/com/epam/tc/hw1/function/multiply/MultiplyFunctionTest.java (hunk 2 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/multiply/MultiplyFunctionTest.java]: Input data types are changed (double -> long). See commit below.

As for methods naming, I would rather leave names as is. The code convention says, that a mothod's name should be a verb (https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html). "Test positive int input" is an action (is a verb). "Positive int input test" or "sumPositiveNumbersTest" is not an action (is a noun).

BTW, I have checked tests on epam autocode service. Test methods have names like: testSingleRoot, test2Roots, etc.

Here is another reason why I wouldn't change the method's name into multiplyPositiveNumbersTest. The mothod is contained by the class MultiplyFunctionTest. I think no need to repeat the function name in the method's name

File: src/test/java/com/epam/tc/hw1/function/multiply/MultiplyFunctionTest.java [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        double actualResult = calculator.mult(multiplicand, multiplier);
+        assertThat(actualResult)
+            .as(String.format(failedTestMessage, "positive integer"))
+            .isEqualTo(expectedResult);
+    }
+
+    @Test(groups = "negative integer input",
+          dataProviderClass = MultiplyFunctionDataProvider.class,
+          dataProvider = "negative integers")
+    public void testNegativeIntInput(double multiplicand, double multiplier, double expectedResult) {
```
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/multiply/MultiplyFunctionTest.java]: Input data types are changed (double -> long). See commit below.

As for methods naming, I would rather leave names as is. The code convention says, that a mothod's name should be a verb (https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html). "Test positive int input" is an action (is a verb). "Positive int input test" or "sumPositiveNumbersTest" is not an action (is a noun).

BTW, I have checked tests on epam autocode service. Test methods have names like: testSingleRoot, test2Roots, etc.

Here is another reason why I wouldn't change the method's name into multiplyPositiveNumbersTest. The mothod is contained by the class MultiplyFunctionTest. I think no need to repeat the function name in the method's name

File: src/test/java/com/epam/tc/hw1/function/multiply/MultiplyFunctionTest.java [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+        double actualResult = calculator.mult(multiplicand, multiplier);
+        assertThat(actualResult)
+            .as(String.format(failedTestMessage, "negative integer"))
+            .isEqualTo(expectedResult);
+    }
+
+    @Test(groups = "mixed integer input",
+          dataProviderClass = MultiplyFunctionDataProvider.class,
+          dataProvider = "mixed integers")
+    public void testMixedIntInput(double multiplicand, double multiplier, double expectedResult) {
```
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/multiply/MultiplyFunctionTest.java]: Input data types are changed (double -> long). See commit below.

As for methods naming, I would rather leave names as is. The code convention says, that a mothod's name should be a verb (https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html). "Test positive int input" is an action (is a verb). "Positive int input test" or "sumPositiveNumbersTest" is not an action (is a noun).

BTW, I have checked tests on epam autocode service. Test methods have names like: testSingleRoot, test2Roots, etc.

Here is another reason why I wouldn't change the method's name into multiplyPositiveNumbersTest. The mothod is contained by the class MultiplyFunctionTest. I think no need to repeat the function name in the method's name

File: src/test/java/com/epam/tc/hw1/function/subtract/SubtractFunctionTest.java (hunk 3 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/subtract/SubtractFunctionTest.java]: Input data types are changed (double -> long). See commit below.

As for methods naming, I would rather leave names as is. The code convention says, that a mothod's name should be a verb (https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html). "Test positive int input" is an action (is a verb). "Positive int input test" or "sumPositiveNumbersTest" is not an action (is a noun).

BTW, I have checked tests on epam autocode service. Test methods have names like: testSingleRoot, test2Roots, etc.

Here is another reason why I wouldn't change the method's name into subtractPositiveNumbersTest. The mothod is contained by the class subtractFunctionTest. I think no need to repeat the function name in the method's name

File: src/test/java/com/epam/tc/hw1/function/subtract/SubtractFunctionTest.java [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+        double actualResult = calculator.sub(minuend, subtrahend);
+        assertThat(actualResult)
+            .as(String.format(failedTestMessage, "positive integer"))
+            .isEqualTo(expectedResult);
+    }
+
+    @Test(groups = "negative integer input",
+          dataProviderClass = SubtractFunctionDataProvider.class,
+          dataProvider = "negative integers")
+    public void testNegativeIntInput(double minuend, double subtrahend, double expectedResult) {
```
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/subtract/SubtractFunctionTest.java]: Input data types are changed (double -> long). See commit below.

As for methods naming, I would rather leave names as is. The code convention says, that a mothod's name should be a verb (https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html). "Test positive int input" is an action (is a verb). "Positive int input test" or "sumPositiveNumbersTest" is not an action (is a noun).

BTW, I have checked tests on epam autocode service. Test methods have names like: testSingleRoot, test2Roots, etc.

Here is another reason why I wouldn't change the method's name into subtractPositiveNumbersTest. The mothod is contained by the class subtractFunctionTest. I think no need to repeat the function name in the method's name

File: src/test/java/com/epam/tc/hw1/function/add/AddFunctionTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/add/AddFunctionTest.java]: Methods renamed.

File: src/test/java/com/epam/tc/hw1/function/add/AddFunctionTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/add/AddFunctionTest.java]: Methods renamed.

File: src/test/java/com/epam/tc/hw1/function/multiply/MultiplyFunctionTest.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/multiply/MultiplyFunctionTest.java]: Methods resolved.

File: src/test/java/com/epam/tc/hw1/function/add/AddFunctionTest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/add/AddFunctionTest.java]: Methods renamed.

File: src/test/java/com/epam/tc/hw1/function/subtract/SubtractFunctionTest.java (hunk 3 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/subtract/SubtractFunctionTest.java]: Methods renamed.

File: src/test/java/com/epam/tc/hw1/function/subtract/SubtractFunctionTest.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/subtract/SubtractFunctionTest.java]: Methods renamed.

File: src/test/java/com/epam/tc/hw1/function/subtract/SubtractFunctionTest.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+        double actualResult = calculator.sub(minuend, subtrahend);
+        assertThat(actualResult)
+            .as(String.format(failedTestMessage, "negative integer"))
+            .isEqualTo(expectedResult);
+    }
+
+    @Test(groups = "mixed integer input",
+          dataProviderClass = SubtractFunctionDataProvider.class,
+          dataProvider = "mixed integers")
+    public void testMixedIntInput(double minuend, double subtrahend, double expectedResult) {
```
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/subtract/SubtractFunctionTest.java]: Methods renamed.

File: src/test/java/com/epam/tc/hw1/function/subtract/SubtractFunctionTest.java (hunk 3 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/subtract/SubtractFunctionTest.java]: Methods renamed.

File: src/test/java/com/epam/tc/hw1/function/subtract/SubtractFunctionTest.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/subtract/SubtractFunctionTest.java]: Methods renamed.

File: src/test/java/com/epam/tc/hw1/function/divide/DivideFunctionTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/divide/DivideFunctionTest.java]: Methods renamed.

File: src/test/java/com/epam/tc/hw1/function/divide/DivideFunctionTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/divide/DivideFunctionTest.java]: Methods renamed.

File: src/test/java/com/epam/tc/hw1/function/multiply/MultiplyFunctionTest.java (hunk 2 above)
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/multiply/MultiplyFunctionTest.java]: Methods renamed.

File: src/test/java/com/epam/tc/hw1/function/multiply/MultiplyFunctionTest.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/multiply/MultiplyFunctionTest.java]: Methods renamed.

File: src/test/java/com/epam/tc/hw1/function/divide/DivideFunctionTest.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (kdr1mkf) [reply src/test/java/com/epam/tc/hw1/function/divide/DivideFunctionTest.java]: Methods renamed.

**Review by ZeroShadows — APPROVED** [review]

**Pull request merged**